Build failure: openmw

### Nixpkgs version

- Unstable (25.11)

### Steps to reproduce

Run `nix build 'git+https://github.com/NixOS/nixpkgs?ref=nixos-unstable#openmw'` and observe the build fails.

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/310421985

### Relevant log output

```console
build/source/components/vfs/qtconversion.cpp: In function 'VFS::Path::Normalized VFS::Path::normalizedFromQString(QStringView)':
/build/source/components/vfs/qtconversion.cpp:19:28: error: call of overloaded 'Normalized(<brace-enclosed initializer list>)' is ambiguous
   19 |     return Normalized{ tmp };
      |                            ^
In file included from /build/source/components/vfs/qtconversion.hpp:6,
                 from /build/source/components/vfs/qtconversion.cpp:2:
/build/source/components/vfs/pathutil.hpp:144:18: note: candidate: 'VFS::Path::Normalized::Normalized(const char*)'
  144 |         explicit Normalized(const char* value)
      |                  ^~~~~~~~~~
/build/source/components/vfs/pathutil.hpp:139:18: note: candidate: 'VFS::Path::Normalized::Normalized(std::string_view)'
  139 |         explicit Normalized(std::string_view value)
      |                  ^~~~~~~~~~
/build/source/components/vfs/qtconversion.cpp: In function 'VFS::Path::Normalized VFS::Path::normalizedFromQString(QString&&)':
/build/source/components/vfs/qtconversion.cpp:25:28: error: call of overloaded 'Normalized(<brace-enclosed initializer list>)' is ambiguous
   25 |     return Normalized{ tmp };
      |                            ^
/build/source/components/vfs/pathutil.hpp:144:18: note: candidate: 'VFS::Path::Normalized::Normalized(const char*)'
  144 |         explicit Normalized(const char* value)
      |                  ^~~~~~~~~~
/build/source/components/vfs/pathutil.hpp:139:18: note: candidate: 'VFS::Path::Normalized::Normalized(std::string_view)'
  139 |         explicit Normalized(std::string_view value)
      |                  ^~~~~~~~~~
[ 47%] Building CXX object apps/essimporter/CMakeFiles/openmw-essimporter.dir/importercontext.cpp.o
[ 47%] Building CXX object apps/essimporter/CMakeFiles/openmw-essimporter.dir/converter.cpp.o
make[2]: *** [components/CMakeFiles/components_qt.dir/build.make:282: components/CMakeFiles/components_qt.dir/vfs/qtconversion.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 47%] Building CXX object apps/openmw/CMakeFiles/openmw-lib.dir/mwrender/navmesh.cpp.o
[ 47%] Building CXX object apps/essimporter/CMakeFiles/openmw-essimporter.dir/convertacdt.cpp.o
[ 47%] Building CXX object apps/essimporter/CMakeFiles/openmw-essimporter.dir/convertnpcc.cpp.o
[ 47%] Building CXX object apps/essimporter/CMakeFiles/openmw-essimporter.dir/convertinventory.cpp.o
[ 47%] Building CXX object apps/openmw/CMakeFiles/openmw-lib.dir/mwrender/actorspaths.cpp.o
[ 47%] Building CXX object apps/essimporter/CMakeFiles/openmw-essimporter.dir/convertcrec.cpp.o
/build/source/components/files/qtconversion.cpp: In function 'std::filesystem::__cxx11::path Files::pathFromQString(QStringView)':
/build/source/components/files/qtconversion.cpp:21:70: error: call of overloaded 'stringToU8String(const QByteArray&)' is ambiguous
   21 |     return std::filesystem::path{ Misc::StringUtils::stringToU8String(tmp) };
      |                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
```

### Additional context

I'm guessing this build failure is due to the bump to Qt 6.10 if that helps.

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.17.3, NixOS, 25.11 (Xantusia), 25.11.20251022.01f116e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2`
 - nixpkgs: `/nix/store/sqilaxdxljyalv530z1qrd1q3nd4lyxq-source`

### Notify maintainers

@marius851000 @SigmaSquadron

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/op/openmw/0001-Do-not-implicitly-convert-QByteArray-to-const-char.patch
+++ b/pkgs/by-name/op/openmw/0001-Do-not-implicitly-convert-QByteArray-to-const-char.patch
@@ -1,0 +1,122 @@
+From 0de42e9a1640e74b2c94ce16ecaf7367ece5f4b0 Mon Sep 17 00:00:00 2001
+From: elsid <elsid.mail@gmail.com>
+Date: Sun, 5 Oct 2025 23:40:17 +0200
+Subject: [PATCH] Do not implicitly convert QByteArray to const char*
+
+Operators supporting this conversion can be disabled via
+QT_NO_CAST_FROM_BYTEARRAY breaking the build. For example:
+
+https://koschei.fedoraproject.org//package/openmw
+https://kojipkgs.fedoraproject.org/work/tasks/5096/137735096/build.log
+
+Backported to OpenMW 0.49.x
+
+Co-authored-by: Sirius902 <10891979+Sirius902@users.noreply.github.com>
+---
+ apps/opencs/model/world/nestedtableproxymodel.cpp |  4 +++-
+ apps/opencs/view/world/util.cpp                   |  5 +++--
+ components/files/qtconversion.cpp                 | 11 ++++++-----
+ components/vfs/qtconversion.cpp                   | 11 ++++++-----
+ 4 files changed, 18 insertions(+), 13 deletions(-)
+
+diff --git a/apps/opencs/model/world/nestedtableproxymodel.cpp b/apps/opencs/model/world/nestedtableproxymodel.cpp
+index f542ce4def..5278710e4a 100644
+--- a/apps/opencs/model/world/nestedtableproxymodel.cpp
++++ b/apps/opencs/model/world/nestedtableproxymodel.cpp
+@@ -13,7 +13,9 @@ CSMWorld::NestedTableProxyModel::NestedTableProxyModel(
+ {
+     const int parentRow = parent.row();
+ 
+-    mId = std::string(parentModel->index(parentRow, 0).data().toString().toUtf8());
++    const QByteArray utf8 = parentModel->index(parentRow, 0).data().toString().toUtf8();
++
++    mId = std::string(utf8.constData(), utf8.size());
+ 
+     QAbstractProxyModel::setSourceModel(parentModel);
+ 
+diff --git a/apps/opencs/view/world/util.cpp b/apps/opencs/view/world/util.cpp
+index dfb587cd96..5fd32d78b4 100644
+--- a/apps/opencs/view/world/util.cpp
++++ b/apps/opencs/view/world/util.cpp
+@@ -363,11 +363,12 @@ void CSVWorld::CommandDelegate::setEditorData(QWidget* editor, const QModelIndex
+     {
+         if (!variant.isValid())
+ #if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
+-            variant = QVariant(editor->property(n).metaType(), (const void*)nullptr);
++            variant = QVariant(editor->property(n.constData()).metaType(), (const void*)nullptr);
++            editor->setProperty(n.constData(), variant);
+ #else
+             variant = QVariant(editor->property(n).userType(), (const void*)nullptr);
++            editor->setProperty(n, variant);
+ #endif
+-        editor->setProperty(n, variant);
+     }
+ }
+ 
+diff --git a/components/files/qtconversion.cpp b/components/files/qtconversion.cpp
+index d56832cb78..ac7ca6ec8b 100644
+--- a/components/files/qtconversion.cpp
++++ b/components/files/qtconversion.cpp
+@@ -1,8 +1,9 @@
+-
+ #include "qtconversion.hpp"
+ 
+ #include <components/misc/strings/conversion.hpp>
+ 
++#include <string_view>
++
+ QString Files::pathToQString(const std::filesystem::path& path)
+ {
+     const auto tmp = path.u8string();
+@@ -17,12 +18,12 @@ QString Files::pathToQString(std::filesystem::path&& path)
+ 
+ std::filesystem::path Files::pathFromQString(QStringView path)
+ {
+-    const auto tmp = path.toUtf8();
+-    return std::filesystem::path{ Misc::StringUtils::stringToU8String(tmp) };
++    const QByteArray tmp = path.toUtf8();
++    return std::filesystem::path(Misc::StringUtils::stringToU8String(std::string_view(tmp.constData(), tmp.size())));
+ }
+ 
+ std::filesystem::path Files::pathFromQString(QString&& path)
+ {
+-    const auto tmp = path.toUtf8();
+-    return std::filesystem::path{ Misc::StringUtils::stringToU8String(tmp) };
++    const QByteArray tmp = path.toUtf8();
++    return std::filesystem::path(Misc::StringUtils::stringToU8String(std::string_view(tmp.constData(), tmp.size())));
+ }
+diff --git a/components/vfs/qtconversion.cpp b/components/vfs/qtconversion.cpp
+index 472e3dd0b4..fc8a0fb78e 100644
+--- a/components/vfs/qtconversion.cpp
++++ b/components/vfs/qtconversion.cpp
+@@ -1,8 +1,9 @@
+-
+ #include "qtconversion.hpp"
+ 
+ #include <components/misc/strings/conversion.hpp>
+ 
++#include <string_view>
++
+ QString VFS::Path::normalizedToQString(NormalizedView path)
+ {
+     return QString::fromUtf8(path.value().data(), path.value().size());
+@@ -15,12 +16,12 @@ QString VFS::Path::normalizedToQString(Normalized&& path)
+ 
+ VFS::Path::Normalized VFS::Path::normalizedFromQString(QStringView path)
+ {
+-    const auto tmp = path.toUtf8();
+-    return Normalized{ tmp };
++    const QByteArray tmp = path.toUtf8();
++    return Normalized(std::string_view(tmp.constData(), tmp.size()));
+ }
+ 
+ VFS::Path::Normalized VFS::Path::normalizedFromQString(QString&& path)
+ {
+-    const auto tmp = path.toUtf8();
+-    return Normalized{ tmp };
++    const QByteArray tmp = std::move(path).toUtf8();
++    return Normalized(std::string_view(tmp.constData(), tmp.size()));
+ }
+-- 
+2.51.0
+

--- a/pkgs/by-name/op/openmw/package.nix
+++ b/pkgs/by-name/op/openmw/package.nix
@@ -85,6 +85,12 @@ stdenv.mkDerivation (finalAttrs: {
     hash = "sha256-Eyjn3jPpo0d7XENg0Ea/3MN60lZBSUAMkz1UtTiIP80=";
   };
 
+  patches = [
+    # https://gitlab.com/OpenMW/openmw/-/merge_requests/4941
+    # FIXME: Remove after next release
+    ./0001-Do-not-implicitly-convert-QByteArray-to-const-char.patch
+  ];
+
   postPatch = ''
     sed '1i#include <memory>' -i components/myguiplatform/myguidatamanager.cpp # gcc12
   ''

